### DISCUSSION TITLE
Ports Overhaul

### DISCUSSION THREAD

**Discussion by 1st1** [geldata/gel#1910] [Pre-RFC]
## Abstract
We propose to simplify network protocol configuration (currently known as "Ports"), add a way to multiplex HTTP protocols and the EdgeDB binary protocol behind one network port, and to implement server-side connection pooling.

## Motivation
Currently, to enable GraphQL over HTTP (or EdgeQL over HTTP) access to EdgeDB, users need to configure an "Application Port":

```
CONFIGURE SYSTEM INSERT Port {
    protocol := "graphql+http",
    database := "your_database_name",
    address := "127.0.0.1",
    port := 8888,
    user := "http",
    concurrency := 4,
};
```

The user would need to specify a network interface for the server to accept connections to the port, as well as to configure the "concurrency" setting.

Specifying a separate network interface for every Application Port is suboptimal. The user needs to make sure to specify an unused network port, as well as configure their application and their network infrastructure to work with the port. If the user wants to configure another Application Port for the same or another DB, they would need to allocate yet another network port. Maintenance of multiple network ports at the same time is inconvenient and needlessly complicates EdgeDB installation, documentation, and tutorials.

The "concurrency" setting is another confusing aspect. The purpose of it is to configure the number of backend PostgreSQL connections to be exclusively reserved for the port. Coming up with the right concurrency number is tricky: if the number is too high, there will be a lot of unused PostgreSQL worker processes, if the number is too low, there won't be enough resources for a good quality of service. Choosing the correct concurrency settings is further complicated if the user wants to run several Application Ports at the same time. That said, the "concurrency" setting effectively implements server-side connection pooling for HTTP protocols in EdgeDB. There is, however, no way to enable server-side pooling for the binary protocol.

Lastly, Application Ports are defined as integral part of the system with unclear boundaries. Our vision is to eventually make EdgeDB a platform where different "extensions" to the system can be installed, supporting different protocols, query dialects, enhancing the type system, having their own custom configuration.

## Specification
In summary, this RFC proposes to:

* Replace the concept of Application Ports with "Extensions".
* Allow the binary protocol and extensions implementing HTTP access points to share the same network port.
* Implement generic server-side connection pooling shared between the binary protocol connections and connections to extensions.

### Extensions

Fully defining the Extension API and how they will be packaged is out of scope of this RFC. That said, we want to define just enough to reclassify the existing "GraphQL over HTTP" and "EdgeQL over HTTP" application ports as "extensions", as well as replace the current `CONFIGURE SYSTEM INSERT Port` with a forward-compatible syntax:

```
  CREATE DATABASE EXTENSION [IF NOT EXISTS] {graphql|edgeql_http} "{"
    subcommand;
  "}";

# where *subcommand* is:

  SET config-option := value;

# where 

  *value* is a string literal;
  *config-option* is an identifier.
```

and

```
DROP DATABASE EXTENSION {graphql|edgeql_http};
```

The above commands allow to enable or disable GraphQL or EdgeQL over HTTP for the current database. Both protocols will be internally hardcoded to work with the new syntax and expose the necessary configuration options, e.g. whether to enable access to GraphiQL UI or not.

Later, the `CREATE SYSTEM EXTENSION` syntax will be allowed to configure system-wide extensions.

The way for the users to interact with configured extensions is detailed in the following section.

### Multiplexing Protocols

This RFC proposes for a single EdgeDB instance to only expose a single network port. All protocols, including the binary protocol, will be multiplexed via it.

When EdgeDB gets support for TLS connections the protocol negotiation will be performed via Application-Layer Protocol Negotiation (ALPN).

Currently though, EdgeDB support only plaintext connections. To make multiplexing of plaintext HTTP and the binary protocol possible, we change the binary protocol to use the HTTP/1.1 upgrade mechanism. Clients, instead of sending a [ClientHandshake](https://www.edgedb.com/docs/internals/protocol/messages#ref-protocol-msg-client-handshake) message, will send an HTTP GET request with an Upgrade header:

```
GET / HTTP/1.1
Host: localhost
Connection: Upgrade
Upgrade: edgedb
```

The server will then reply with:

```
HTTP/1.1 101 Switching Protocols
Upgrade: edgedb
Connection: Upgrade
```

After receiving which, the client must send the `ClientHandshake` binary protocol message.

The GraphQL and EdgeQL/HTTP extensions will also be accessible via the same port and the HTTP protocol. The following URL structure is adapted:

```
/db/{DATABASE}/{EXTENSION}/
```

The proposed URL structure allows multiple different extensions to expose their HTTP endpoints for different databases and even at the system level.

For example, if the user enables the "graphql" extension on their "test" database on their development machine, the GraphQL endpoint could be accessed via the following URLs:

* Main GraphQL endpoint: `http://localhost:5656/db/test/graphql`
* GraphiQL UI: `http://localhost:5656/db/test/graphql/explore`

### Connection Pooling

EdgeDB will expose two new system configuration options:

* `max_backend_connections`: the maximum number of simultaneously open PostgreSQL connections that EdgeDB can use.
* `reserved_superuser_connections`: the number of connections exempt from the connection pooling and reserved for superusers and administrators connecting to the database.

Effectively, the server will use up to `max_backend_connections - reserved_superuser_connections` PostgreSQL connections to serve the binary protocol and HTTP extensions. If a superuser connects to the DB and there's no connection available in the pool immediately, a special reserved PostgreSQL connection would be pinned to their EdgeDB session.

### Connection Pooling

One of the complications of implementing the server-side connection pooling for EdgeDB is the fact that PostgreSQL workers cannot switch to a different database. And starting a new worker is relatively expensive. Therefore EdgeDB will need to implement some form of Weighted Fair Queuing algorithm to automatically balance the pool of available connections for every accessed database of the instance. The good news is that the majority of users typically use just one database, so our scheduling algorithms do not need to be perfect in 1.0.

Another complication is scaling the compiler processes to be able to compile queries for different databases. Currently every compiler OS process is pinned to one database and is responsible for introspecting the schema. This will have to change: the server has to manage the state of all open connections (and their transient schema states) and be able to push that state to the compiler processes. The preliminary work of separating the frozen "std" schema from the "user" schema [has already been completed](https://github.com/edgedb/edgedb/pull/1884).

## Backwards Compatibility
To maintain backwards compatibility until 1.0 is released, the server will still accept binary connections that send a `ClientHandshake` message instead of an HTTP Upgrade request. The `ClientHandshake` message starts with the 0x56 byte ("V" in ASCII) and the server can detect that and start parsing the binary protocol right away. The compatibility will be dropped before 1.0 final is released. All official EdgeDB bindings and CLI tools will be updated to use the HTTP upgrade mechanism. The CLI can maintain backwards compatibility separately; if the server replies with an `ErrorResponse` binary message to the HTTP upgrade request it would mean that the server is pre-1.0 and does not support HTTP upgrade.

`CONFIGURE SYSTEM INSERT Port` will also be supported up until the 1.0 release.

**Comment by tailhook** [comment]
1. Database name should be encoded as `EncodeURIComponent` (or whatever) in the url
2. `reserved_superuser_connections` is useless for now as well as connection pinning as all our users are superusers now. I propose to postpone this
3. I think that we need database configuration setting `reserved_backend_connections` that reserves connections for this database. So that in case our scheduling algorithm creates a lot of connection churn there is a way to fix it quickly without modifying database code.
4. There is no description of how connections are allocated:
     a) Explicit transaction obviously holds the connection
     b) Is connection allocated when doing `Prepare`? Is it deallocated during `Sync`? What if there is good old `Prepare, Sync, Execute, Sync` cycle? What if client decided not to run `Execute` after prepare?
5. There is no mention of **replica** connections. We expect to have at least some support of them at 1.0. And they need separate set of tunables (and probably some per-replica settings). And perhaps even different connection allocation strategy.
6. If we go with `CREATE EXTENSION` (but see below) the following should be mentioned:
    a) `ALTER EXTENSION` (changing config)
    b) Figure out a way to warn users about new experimental extensions, like `CREATE EXPERIMENTAL EXTENSION` or `CREATE EXTENSION forms_beta`
    c) Mention it is dumped within database dump (or doesn't it? Optionally?) This is important for experimental extension lifecycle.

----

I don't think `CREATE EXTENSION` is good enough. It's there in postgres to load `.so` file (some of them can be dangerous to load by default). But for us:
1. Our built-in extensions aren't going to be `.so` files and are not loaded on per-database basis.
2. I expect our pluggable extensions are going to be `.wasm`, so they are safe to load.
3. But since we have cloud, we're probably want uploading `.wasm` files by the protocol for user defined extensions (so out of scope for this RFC).
4. We may want payed extensions in the cloud. So it's dangerous to use `CREATE EXTENSION` for them, and we'd rather have them enabled in admin panel UI and via admin API (i.e. API that can be used to create a new edgedb instance).


So what I propose instead:
1. `edgedb-server --all-extensions/--disable-extensions/--experimental-extensions/--extensions=xxx,yyy,zzz` (perhaps `--all-extensions` that doesn't include experimental ones should be default)
2. For cloud we obviously have some runtime tunables but this is out of scope of this protocol/RFC
3. Layered configuration: `CONFIGURE SYSTEM SET graphql_config := { enable := false };` and `CONFIGURE DATABASE SET graphql_config := { enable := true };`. The pros are:
    a) It uses already available constructions and is dumped as a normal part of the configuration
    b) configuration can be set globally and can be overridden on per-database basis
    (but we should figure out how to handle restore to the server that doesn't have extension available)

**Reply by 1st1** [reply]
> Database name should be encoded as EncodeURIComponent (or whatever) in the url

Sure, as everything else (extension name).

> reserved_superuser_connections is useless for now as well as connection pinning as all our users are superusers now. I propose to postpone this

"as all our users are superusers now" @elprans do we want to fix this before 1.0?

> I think that we need database configuration setting reserved_backend_connections that reserves connections for this database. So that in case our scheduling algorithm creates a lot of connection churn there is a way to fix it quickly without modifying database code.

I like this. And we now have infrastructure for this.

> There is no description of how connections are allocated:
a) Explicit transaction obviously holds the connection
b) Is connection allocated when doing Prepare? Is it deallocated during Sync? What if there is good old Prepare, Sync, Execute, Sync cycle? What if client decided not to run Execute after prepare?

This is a bit of an implementation detail, but we can clarify it in the RFC (maybe once we start implemeting it?)  The rule of thumb is that we only take a backend connection from the pool when we actually need it. Mostly it's when we're running a compiled query (Execute message).

> Is it deallocated during Sync?

Yes. Because a transaction is created between Syncs. We'll need to document that the client needs to send Sync as soon as possible (ideally right with the Execute and other packets).

> What if client decided not to run Execute after prepare?

The connection will be acquired lazily, e.g.:

The client sends: `[Prepare][Flush]`; the server compiles the query, no need for a postgres connection. The client sends `[Execute]`; the server acquires a backend connection to run the query. The client sends `[Sync]` - the server executes Sync and releases the connection. We'll document that good clients should group `[Execute][Sync]` in one message; that way the server won't hold the connection longer than necessary.

> There is no mention of replica connections. We expect to have at least some support of them at 1.0. And they need separate set of tunables (and probably some per-replica settings). And perhaps even different connection allocation strategy.

I think this is slightly out of scope of this RFC, but I'm not against adding clarifications and more details if you want to contribute them. Ultimately, the RFC outlines the big picture here. We'll be able to further evolve the spec and add more config parameters as we go.

---------

> I don't think CREATE EXTENSION is good enough. It's there in postgres to load .so file (some of them can be dangerous to load by default). But for us:

>    Our built-in extensions aren't going to be .so files and are not loaded on per-database basis.

Right, but this is an implementation detail of 1.0. I actually do want to separate the GraphQL extension into a separate repo and maybe it will be written in Rust and be a .so or .wasm file in some EdgeDB X.0.

> I expect our pluggable extensions are going to be .wasm, so they are safe to load.
>But since we have cloud, we're probably want uploading .wasm files by the protocol for user defined extensions (so out of scope for this RFC).
>We may want payed extensions in the cloud. So it's dangerous to use CREATE EXTENSION for them, and we'd rather have them enabled in admin panel UI and via admin API (i.e. API that can be used to create a new edgedb instance).

Hm, I don't see how any of that is an argument against CREATE EXTENSION. Paid extensions can throw an error if you try to CREATE EXTENSION them and advise to use the cloud panel.

> Layered configuration: CONFIGURE SYSTEM SET graphql_config := { enable := false }; and CONFIGURE DATABASE SET graphql_config := { enable := true };. The pros are:

The whole point of this proposal is to abstract away the fact that GraphQL is a built-in thing and mock it as if it was a separate extension (as it will be in the future). So I'm -1 on using broad system settings to configure it.

**Reply by 1st1** [reply]
We can also tune the `Sync` implementation to only acquire the connection when needed. E.g. if the client sends `[Prepare][Sync]` we don't need to acquire one, because there's no open implicit transaction yet.

**Reply by tailhook** [reply]
> This is a bit of an implementation detail, but we can clarify it in the RFC

This is as important info for anyone implementing the protocol as the message format. So yes it has to be documented and RFC is a good place.

> The client sends: [Prepare][Flush]; the server compiles the query, no need for a postgres connection. The client sends [Execute]; the server acquires a backend connection to run the query.

This is inefficient as we need to re-prepare postgres query in the new connection. And I think we need postgres connection for prepare to describe set of arguments (and we need it before describe to return a typeid hash). Unless statement is already in cache. So not sure about the right trade-off here. That's one reason why the decision with pros and cons should be in the RFC.

>>   There is no mention of replica connections. We expect to have at least some support of them at 1.0. And they need separate set of tunables (and probably some per-replica settings). And perhaps even different connection allocation strategy.
> 
> I think this is slightly out of scope of this RFC, but I'm not against adding clarifications and more details if you want to contribute them. Ultimately, the RFC outlines the big picture here. We'll be able to further evolve the spec and add more config parameters as we go.

It would be nice to discuss together so configuration of master and slaves are not very different from each other.

> Right, but this is an implementation detail of 1.0. I actually do want to separate the GraphQL extension into a separate repo and maybe it will be written in Rust and be a .so or .wasm file in some EdgeDB X.0.

I still don't think there is any need to "create". Anything that is built-in can be loaded lazily (when enabled by config). If we don't have `wasm` file at the server then `CREATE` can't program `wasm` for us. Also see below.

> Hm, I don't see how any of that is an argument against CREATE EXTENSION. Paid extensions can throw an error if you try to CREATE EXTENSION them and advise to use the cloud panel.

This is also true, but why you have to `CREATE` something that you have already enabled in UI?

> The whole point of this proposal is to abstract away the fact that GraphQL is a built-in thing and mock it as if it was a separate extension (as it will be in the future). So I'm -1 on using broad system settings to configure it.

I don't understand what you mean by "broad". My proposal doesn't say that there should be difference between builtin and external extensions, what it says:
1. Extensions are basically per-server, built-ins are already there. If we upload a wasm, there is no reason to upload it to each database and store it in each database's dump (if we ever have wasm functions, then yes, but that's different from extensions). I'm not even sure that uploaded dump should be a part of `dump --all` because that makes upgrading extensions harder.
2. Config can be loaded lazily if you access `graphql_config` and there is extension named `graphql` that is not loaded: load it and proceed with config. Or we can put it into an `extensions { graphql }`. Structure is TBD.

Another upside of the `CONFIG` is that when scripting you don't need to find out whether you use `CREATE` or `ALTER`. You just put the target configuration an that just works. Or do we plan to include `CREATE EXTENSION` into migration files? (this has to be mentioned in the RFC then).

**Reply by elprans** [reply]
> I don't think `CREATE EXTENSION` is good enough. It's there in postgres to load .so file (some of them can be dangerous to load by default)

Not exactly. _Some_ extensions are implemented in C, but `CREATE EXTENSION` is conceptually just a wrapper around an SQL script + some metadata.  Our implementation is likely going to be similar: `CREATE EXTENSION` would run some DDL.  The underlying system objects, like the protocol registration are likely going to use a dedicated syntax (`CREATE PROTOCOL`?).  The extension script can also register configuration types and issue `CONFIGURE` as necessary, so configuration persistence is not an issue in this design.  Obviously, none of this needs to be fully fleshed out now, we just need to define the high-level API for now.

**Reply by 1st1** [reply]
@tailhook:

>>    The client sends: [Prepare][Flush]; the server compiles the query, no need for a postgres connection. The client sends [Execute]; the server acquires a backend connection to run the query.

>This is inefficient as we need to re-prepare postgres query in the new connection. 

We need to re-prepare but we only need to prepare right before the execute. We know everything about PostgreSQL types and EdgeQL types after we compile the query. IOW EdgeDB's [prepare] does not need a postgres connection.

> And I think we need postgres connection for prepare to describe set of arguments (and we need it before describe to return a typeid hash). 

No (unless I'm missing something new).

> Unless statement is already in cache. So not sure about the right trade-off here. That's one reason why the decision with pros and cons should be in the RFC.

"with pros and cons" of what? I don't see any alternative and so far I honestly don't think there's a problem here.

-----



>>>        There is no mention of replica connections. We expect to have at least some support of them at 1.0. And they need separate set of tunables (and probably some per-replica settings). And perhaps even different connection allocation strategy.

>>    I think this is slightly out of scope of this RFC, but I'm not against adding clarifications and more details if you want to contribute them. Ultimately, the RFC outlines the big picture here. We'll be able to further evolve the spec and add more config parameters as we go.

> It would be nice to discuss together so configuration of master and slaves are not very different from each other.

You mean discuss how replica's parameters like `max_backend_connections` would be configured? Should they even be configured via the leader instance? @elprans 

-----


>>    Hm, I don't see how any of that is an argument against CREATE EXTENSION. Paid extensions can throw an error if you try to CREATE EXTENSION them and advise to use the cloud panel.

> This is also true, but why you have to CREATE something that you have already enabled in UI?

You wouldn't. Once you enable a paid feature we can just make an entry to some system table in the same way `CREATE EXTENSION` would.

----

> I don't understand what you mean by "broad". My proposal doesn't say that there should be difference between builtin and external extensions, what it says:

I guess I was confused by what you meant with `CONFIGURE SYSTEM SET graphql_config := { enable := false };`. I thought it was some hardcoded option to configure the built-in graphql implementation. The way I understand it now is that `CONFIGURE SYSTEM SET XXX_config` would be a way configure an extension named `XXX`, is that right?

-----

@elprans 



>>    I don't think CREATE EXTENSION is good enough. It's there in postgres to load .so file (some of them can be dangerous to load by default)

> Not exactly. Some extensions are implemented in C, but CREATE EXTENSION is conceptually just a wrapper around an SQL script + some metadata. Our implementation is likely going to be similar: CREATE EXTENSION would run some DDL. The underlying system objects, like the protocol registration are likely going to use a dedicated syntax (CREATE PROTOCOL?). The extension script can also register configuration types and issue CONFIGURE as necessary, so configuration persistence is not an issue in this design. Obviously, none of this needs to be fully fleshed out now, we just need to define the high-level API for now.

Yeah, exactly. Like our GIS extension will likely be a collection of files, like I imagine: `schema.esdl` to describe all EdgeQL types/functions/operators and map them to their corresponding PostGIS SQL types; `setup.edgeql` to bootstrap what cannot be specified via ESDL; `manifest.json` and the PostGIS' `.so` extension files.  EdgeQL's `CREATE EXTENSION gis` would translate to SQL `CREATE EXTENSION postgis` and a bunch of other SQL to take care of schema bootstrap.

**Reply by tailhook** [reply]
> "with pros and cons" of what? I don't see any alternative and so far I honestly don't think there's a problem here.

Okay. I assumed that we prepare postgres statements when doing our prepare. Saying that we don't do that might still belong to this RFC as connection pooling makes this extremely more complex to implement (or explain how we can achieve this in the future updates). But theoretically that's how you get the most of postgres performance, right? (i know this is rarely done in practice, as everyone uses pgbouncer, but improving things is what we do within this project).

That said, we [sometimes recompile](https://github.com/edgedb/edgedb/blob/6c5411e083189d40b80b1859a35701837e45b18f/edb/server/mng_port/edgecon.pyx#L851-L852) queries on Execute, but not sure if connection pooling changes anything there. 

> You mean discuss how replica's parameters like max_backend_connections would be configured? Should they even be configured via the leader instance? @elprans

Yes. I think we agreed that same EdgeDB instance will be connected both to master and replicas (at least this is what assumed in #1859 ). Then the read-only instance of the EdgeDB itself is post 1.0 material (as it requires schema synchronization).

> The way I understand it now is that CONFIGURE SYSTEM SET XXX_config would be a way configure an extension named XXX, is that right?

Yes.

> Not exactly. Some extensions are implemented in C, but CREATE EXTENSION is conceptually just a wrapper around an SQL script + some metadata. Our implementation is likely going to be similar: CREATE EXTENSION would run some DDL.

Then I believe that extensions should be a part of SDL, so it is present in **schema** and **migrations**. Because user defined types depend on them. (And then yes, in DDL it could look like `CREATE EXTENSION` but that's less important than SDL syntax). And then runtime configuration (i.e. connection limit) is a part of configuration system.

**Reply by 1st1** [reply]
> That said, we sometimes recompile queries on Execute, but not sure if connection pooling changes anything there.

I don't think it does.

> Yes. I think we agreed that same EdgeDB instance will be connected both to master and replicas (at least this is what assumed in #1859 ). Then the read-only instance of the EdgeDB itself is post 1.0 material (as it requires schema synchronization).

I'll wait for @elprans to chime in on this.

>> The way I understand it now is that CONFIGURE SYSTEM SET XXX_config would be a way configure an extension named XXX, is that right?
>
> Yes.

I agree and it's a good idea. I'll update the RFC to mention configuration, I have a few ideas.

> Then I believe that extensions should be a part of SDL, so it is present in schema and migrations. Because user defined types depend on them. (And then yes, in DDL it could look like CREATE EXTENSION but that's less important than SDL syntax)

Right, I can see that. But I don't think we need that in 1.0, specifically because we don't have the bandwidth right now to define that inner structure of extensions (and understand what we'll have in the SDL). So far I can only imagine this:

```
database extension graphql {
  graphiql := true;
}
```

> And then runtime configuration (i.e. connection limit) is a part of configuration system.

Can you clarify what you meant with "connection limit" being configurable? What that config knob will do?

**Reply by tailhook** [reply]
> So far I can only imagine this:
> 
> ```
> database extension graphql {
>   graphiql := true;
> }
> ```
> 

Well, `graphiql` should be runtime configuration not a part of SDL. Because this is something should be enabled on dev environment but perhaps disabled on production.

What should be configured in SDL is anything that changes public types of the extensions. Or perhaps access control, i.e.:
```
database extension graphql {
  exposed_types := {'User', 'Article'};
}
```

> > And then runtime configuration (i.e. connection limit) is a part of configuration system.
> 
> Can you clarify what you meant with "connection limit" being configurable? What that config knob will do?

Something along the lines of:
```
CONFIGURE SYSTEM SET graphql_config := {
    graphiql := True,
    max_http_connections := 1000,
    keep_alive_timeout := 100,
};
```
The basic idea is if we have ulimit of `3000`, we can split it between 1000 http connections and 1000 binary proto connections and keep 1000 descriptors spare for backend connections and other auxiliary needs.

**Reply by 1st1** [reply]
> What should be configured in SDL is anything that changes public types of the extensions. Or perhaps access control, i.e.:
>
> database extension graphql {
>  exposed_types := {'User', 'Article'};
> }

We plan to have database views for this, eventually. It's an open question how those things will behave, but most likely you'll be "creating database extension for the view". And then the view would only export some of the types with some of the pointers and with some additional ACL rules.

**Reply by 1st1** [reply]
> The basic idea is if we have ulimit of 3000, we can split it between 1000 http connections and 1000 binary proto connections and keep 1000 descriptors spare for backend connections and other auxiliary needs.

Are you sure we should bother with this at the EdgeDB level? IMO things like this should be configured at some HTTP balancer layer in your cloud.

**Reply by tailhook** [reply]
Well, generally both `exposed_types` and `max_http_connections` were just examples of different things that belong to different places. But...

>> The basic idea is if we have ulimit of 3000, we can split it between 1000 http connections and 1000 binary proto connections and keep 1000 descriptors spare for backend connections and other auxiliary needs.
>
> Are you sure we should bother with this at the EdgeDB level? IMO things like this should be configured at some HTTP balancer layer in your cloud.

Yes. Because:
1. At least nginx can't control number of backend connections (I don't know if others can)
2. If we don't limit number of incoming connections and exhaust ulimit, we get lots of ugly errors in the middle of otherwise valid code, that includes errors and crashes in older connections. I.e. any kind of load spike or misconfiguration or DoS is a nightmare.

That said we might limit only total number of connections for both HTTP and TCP proto, if we don't care.

> We plan to have database views for this, eventually. It's an open question how those things will behave, but most likely you'll be "creating database extension for the view". And then the view would only export some of the types with some of the pointers and with some additional ACL rules.

Well, I've though View == Alias. I.e. it exposes subset of type, but not subset of the database. But anyway, this example might be bad, my point was to show that both SDL settings and runtime configs make sense for extensions.

**Reply by 1st1** [reply]
> That said we might limit only total number of connections for both HTTP and TCP proto, if we don't care.

I think the total number of connections is a much more reasonable setting. And it's easier to implement/configure for us. I'll include it to the RFC.

**Comment by tailhook** [comment]
Also let's consider the two more ports/extensions that we should around 1.0:
1. Healthchecks
2. Metrics

Both of them:
1. Do not introduce any new types
2. Must be toggled as part of config not SDL, preferably enabled by default

Or aren't they extensions?

**Reply by 1st1** [reply]
I'd say they should be extensions just so that we have one unified mechanism for enabling/disabling things.

**Reply by elprans** [reply]
The split between configuration-only and schema-affecting extensions is bothering me somewhat.  We need to better define the configuration portability (i.e. what goes into dump by default and what doesn't), and, relatedly, what should appear in SDL and what shouldn't.

**Comment by tailhook** [comment]
I don't know about 'allowed_types', but what definitely has to be in `graphql` config is some sort of `readonly := true`. And this is quite complex question if it has to be SDL setting or database setting.

**Comment by 1st1** [comment]
We also need to make CORS configurable.

**Reply by tailhook** [reply]
CORS should probably be a database-level setting?

**Reply by 1st1** [reply]
Per-database and per-extension, I think.

**Comment by elprans** [comment]
> ```
>   CREATE DATABASE EXTENSION [IF NOT EXISTS] {graphql|edgeql_http} "{"
>     subcommand;
>   "}";
> 
> # where *subcommand* is:
> 
>   SET config-option := value;
> 
> # where 
> 
>   *value* is a string literal;
>   *config-option* is an identifier.
> ```

I've re-read the discussion and I'm leaning toward agreement with @tailhook with respect to how extendion "config" should be handled.  Fundamentally `CREATE EXTENSION` creates a schema object, and so any subcommands in the statement should be about how the extension affects the schema, and *not* about its runtime configuration.  For example:

```
CREATE DATABASE EXTENSION gis {
    SET topology_spatial_types := true;
};
```

In other words, `CREATE EXTENSION` is like calling a DDL "macro" with parameters.  I even briefly considered function-call syntax, but that would make `ALTER` awkward.

In my opinion, runtime configuration of extensions should be performed with `CONFIGURE`.  For example:

```
CREATE DATABASE EXTENSION graphql;
CONFIGURE DATABASE SET graphql::cors := '*';
```

**Reply by 1st1** [reply]
I really don't like the namespace-like extension hack `graphql::cors`. `CREATE DATABASE EXTENSION graphql;` should not equate to `CREATE MODULE graphql`. 

IMO it should simply be `graphql_cors` or `graphql__cors` (to make it easier to tell apart config settings from extensions that have `_` in their name).

**Reply by elprans** [reply]
> IMO it should simply be `graphql_cors` or `graphql__cors` (to make it easier to tell apart config settings from extensions that have `_` in their name).

How would that work, given that `cfg::` is read-only and you must declare the configuration parameter in the schema before it can be used?

**Reply by tailhook** [reply]
> I really don't like the namespace-like extension hack `graphql::cors`. `CREATE DATABASE EXTENSION graphql;` should not equate to `CREATE MODULE graphql`.
> 

Are you concerned with name conflict with user namespaces? We can call it clearer, like `graphql_config::cors` or `graphql_extension::cors`.

**Reply by tailhook** [reply]
> How would that work, given that `cfg::` is read-only and you must declare the configuration parameter in the schema before it can be used?

I'm not sure how hard the change it would be. While I agree that directly altering `cfg::` is not what is expected to work. But I don't see why we can't use some extension's metadata to add settings internally.

**Reply by elprans** [reply]
I'd really like to avoid inventing new systems here.  There is already a functioning and tested config system that relies on DDL declarations.  Extensions are supposed to essentially be addressable "DDL containers", so I don't see why proper config type declarations shouldn't be part of it.

**Comment by 1st1** [comment]
Had a quick discussion about this on our sync call:

* `DECLARE EXTENSION { ... }` syntax to define what's in the extension. An extension can define config options, modules, etc.

* Our `stdgraphql` module will be renamed to `graphql` and will be part of the "graphql" extension.

* `CREATE EXTENSION` and `DROP EXTENSION` will be used to install/uninstall extension to a particular DB.

I propose to limit ourselves to declaring just one module per extension for now. Once we have nested namespaces, it will be more convenient to reason about an extension if it exports just one module, not an array of top-level modules.

**Reply by elprans** [reply]
Two more operations are needed, albeit not necessarily right now:

- `UPDATE EXTENSION` to "migrate" an existing installed extension to a new version
- Related, a way to re-`DECLARE` an extension using an updated source (possibly just allow `DECLARE EXTENSION` to overwrite, or add a modifier, like `DECLARE OR REPLACE EXTENSION`)

**Reply by 1st1** [reply]
> `DECLARE OR REPLACE EXTENSION`

`DECLARE EXTENSION IF NOT EXISTS` ?

**Reply by elprans** [reply]
> `DECLARE EXTENSION IF NOT EXISTS`

No, this is conditional declaration.  What if I want to _replace_ the existing extension?

**Reply by 1st1** [reply]
> What if I want to replace the existing extension?

I'm not sure what replacing means. Can you elaborate on specifics? What will it do to the existing config? Can I replace an extension with another completely unrelated one sharing the same name? etc.

**Reply by elprans** [reply]
> I'm not sure what replacing means.

`DECLARE EXTENSION` stores extension declaration in the database.  Consider a third-party GIS extension.  You declare an initial version of extension downloaded from Github or something.  `DECLARE EXTENSION gis { ... }` (we will likely have a formal CLI command to do this or maybe even a "package manager").  Then, upstream releases an update for the extension.  And so, you want to repeat the `DECLARE` step to get the updated declaration recorded. There are two options: 1) `DECLARE EXTENSION` replaces the declaration unconditionally, or 2) we add an explicit qualifier to replace the declaration.

> What will it do to the existing config?

Nothing.  `DECLARE` only replaces the declaration.  To upgrade the actual installed extension you must do `UPDATE EXTENSION`, which will presumably trigger the upgrade DDL declared in the extension (Postgres does this with SQL files containing DDL: [example](https://github.com/postgres/postgres/blob/master/contrib/cube/cube--1.2--1.3.sql)).

> Can I replace an extension with another completely unrelated one sharing the same name?

Yes, but why would you want to?

**Reply by 1st1** [reply]
>  1) DECLARE EXTENSION replaces the declaration unconditionally, or 2) we add an explicit qualifier to replace the declaration.

Will I be able to just "undeclare" it, without replacing?

>> Can I replace an extension with another completely unrelated one sharing the same name?
>
> Yes, but why would you want to?

By accident.

**Reply by elprans** [reply]
> Will I be able to just "undeclare" it, without replacing?

We will need a keyword for this.

> By accident.

I'm not sure this accident is likely, given that `DECLARE EXTENSION` isn't something you'd be typing in REPL.

**Reply by 1st1** [reply]
Well, it's more about having two conflicting packages -- it's unclear what's going on in cases like that. But I guess just like in, say, Python, it will be up to the user to figure out this mess. We could require all extensions to have a unique ID annotation in them, which would at least make it possible for us to handle collisions better.

**Reply by elprans** [reply]
Alternatively to `DECLARE EXTENSION` we can use `CREATE EXTENSION DECLARATION`, which would naturally allow `ALTER EXTENSION DECLARATION` and `DROP EXTENSION DECLARATION` and would align better with the underlying object structure, as we would have to have an `ExtensionDeclaration` class, which would likely be a global object, and an `Extension` class, which would represent registered extensions produced by executing an `ExtensionDeclaration`.

**Reply by elprans** [reply]
Also, instead of `DECLARATION` we can use `PACKAGE`

**Reply by tailhook** [reply]
My thoughts are:
1. Extension could provide types or functions
2. Then schema might depend on that types and/or functions
3. Then we have to have SDL, that handles that so:
    ```
    extension graphql {
        version := "1.1";
    }
    ```
4. So updating extension is something like `ALTER EXTENSION graphql { update to version := "1.1"; }`
